### PR TITLE
fix multi clusters proxies with same ip

### PR DESCRIPTION
**Pull request by hzxuzhonghu**

### CODE REVIEW

**Comment by fpesce** [comment]
Where is the corresponding master commit/PR exactly?
c.f. point 5 [1.2 Release branch cut](https://discuss.istio.io/t/1-2-release-branch-cut/2372)

**Comment by hzxuzhonghu** [comment]
@fpesce #14072

**Comment by howardjohn** [comment]
Like I mentioned on https://github.com/istio/istio/pull/14072, needs the flag that was in the 1.1 commit.

**Comment by fpesce** [comment]
> Like I mentioned on #14072, needs the flag that was in the 1.1 commit.

@hzxuzhonghu So, what is correct move here? should we have a flag like in the 1.1 commit?

**Comment by hzxuzhonghu** [comment]
@fpesce @howardjohn I would rather waiting some days until we release 1.2 than adding the flag.

**Comment by andraxylia** [comment]
Can you please add description to the PR indicating the problem it solves and how? This will help the release manager in their decision.

File: pilot/pkg/serviceregistry/kube/controller.go [hunk 1]
```diff
@@ -491,6 +491,14 @@ func (c *Controller) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.Serv
 
 	pod := c.pods.getPodByIP(proxyIP)
 	if pod != nil {
+		// for split horizon EDS k8s multi cluster, in case there are pods of the same ip across clusters,
+		// which can happen when multi clusters using same pod cidr.
+		// As we have proxy Network meta, compare it with the network which endpoint belongs to,
+		// if they are not same, ignore the pod, because the pod is in another cluster.
+		if proxy.Metadata[model.NodeMetadataNetwork] != c.endpointNetwork(proxyIP) {
```
Comment (rshriram) [inline pilot/pkg/serviceregistry/kube/controller.go]: does this default to "" when the network is not set on proxies (in single cluster mode)?

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller.go]: I think this is a valid concern. What happens if my proxy is missing the network metadata (or alternatively endpointNetwork returns "" -- looking at the code this seems possible). In either one of these cases we could run into this code on single cluster.

Maybe we need to do this check only if both are `!= ""`? not sure if there is a better solution

**Comment by fpesce** [comment]
After discussion with @andraxylia I'd like to not block this, as it seems to be a very straightforward and simple fix.

I've heard @nmittler is an expert in the multiclusters domain, Nathan, if you give a lgtm to that, I'll merge to 1.2.
Thanks everyone for jumping in and clarifying this to me.

**Comment by andraxylia** [comment]
@rshriram @nmittler what is the plan for having an automated test for multi-cluster?

**Comment by nmittler** [comment]
> @rshriram @nmittler what is the plan for having an automated test for multi-cluster?

@andraxylia it depends on what you mean exactly. If you're asking for a test for @Nino-K's work on synthetic service entries, he's going to start looking at that now. Apart from that, I'm not aware of any existing plans. It's something that we should start looking at soon, however.

File: pilot/pkg/serviceregistry/kube/controller.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller.go]: By default it is "" for both proxy and endpoint network.

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller.go]: I see, thanks, so we shouldn't have this issue in single cluster

**Pull request merged**